File: the_money_example/c++/main/money.cpp
#include "money.h"

Money::Money(int amount, string currency) {
  this->amount = amount;
  this->_currency = currency;
}

string Money::currency() const { return _currency; }

Money Money::times(int multiplier) { return Money(amount * multiplier, _currency); }

bool Money::operator==(Money money) const {
  return amount == money.amount && _currency == money._currency;
}

Money Money::operator+(Money addend) {
  return Money(amount + addend.amount, _currency);
}

Money dollar(int amount) {
  return Money(amount, "USD");
}

Money franc(int amount) {
  return Money(amount, "CHF");
}

File: the_money_example/csharp/TheMoneyExample/Money.cs
namespace Money;

public class Money : IExpression
{
    public int Amount { get; }

    public String Currency { get; }

    public Money(int amount, String currency)
    {
        Amount = amount;
        Currency = currency;
    }

    public Money Times(int multiplier)
    {
        return new Money(Amount * multiplier, Currency);

    }

    public IExpression Plus(Money addend)
    {
        return new Sum(this, addend);
    }

    public Money Reduce(String to)
    {
        return this;
    }

    public override bool Equals(Object? obj)
    {
        Money money = (Money)obj!;
        return Amount == money.Amount && Currency.Equals(money.Currency);
    }

    public override string ToString()
    {
        return $"{Amount} {Currency}";
    }

    public static Money Dollar(int amount)
    {
        return new Money(amount, "USD");
    }

    public static Money Franc(int amount)
    {
        return new Money(amount, "CHF");
    }
}
